Implement Observable#inspect

maraisr opened the pull request:
#### 8b80739ea60b5d9b0a7cd7a9152c4e5b15ff8137
<pre>
Implement Observable#inspect
<a href="https://bugs.webkit.org/show_bug.cgi?id=282083">https://bugs.webkit.org/show_bug.cgi?id=282083</a>

Reviewed by Chris Dumez.

This change introduces the .inspect operator for Observables, using the
InternalObserver, and a new SubscriberCallbackInspect. .inspect is meant to act
as a pass-through operator, so there are no mutations on events.

Spec: &lt;<a href="https://wicg.github.io/observable/#dom-observable-inspect">https://wicg.github.io/observable/#dom-observable-inspect</a>&gt;

There is also a new ObservableInspector IDL dictionary created, as well as a
ObservableInspectorAbortCallback. We could not just have used some generic
callback IDL, given we require a callback that accepts a JSC::Value, as we
forward abort reasons to this callback.

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-inspect.any-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-inspect.any.worker-expected.txt: Rebaseline.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/InternalObserverInspect.cpp: Added.
(WebCore::createSubscriberCallbackInspect):
* Source/WebCore/dom/InternalObserverInspect.h: Added.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::inspect):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:
Exposes a `Observable inspect(inspectorUnion)` operator.
* Source/WebCore/dom/ObservableInspector.h: Added.
* Source/WebCore/dom/ObservableInspector.idl: Added.
* Source/WebCore/dom/ObservableInspectorAbortCallback.h: Added.
* Source/WebCore/dom/ObservableInspectorAbortCallback.idl: Added.

Canonical link: <a href="https://commits.webkit.org/287504@main">https://commits.webkit.org/287504@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5050ffd1e0605dd9a9a120371e81ef749630e7a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79889 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58887 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33287 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84404 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30869 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81998 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67951 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7179 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62432 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20264 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82957 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52515 "Found 1 new test failure: media/audioSession/getUserMedia.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72753 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42743 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49831 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26902 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29326 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70952 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27379 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85836 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7107 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4980 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70705 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7284 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68593 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69947 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13947 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12875 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12361 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7071 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12620 "Found 1 new failure in JSWebCodecsEncodedVideoChunk.cpp") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6921 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10432 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8726 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

maraisr commented on LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-inspect.any-expected.txt:
```diff
@@ -1,79 +1,16 @@
+PASS inspect(): Errors thrown in subscribe() Inspector handler subscribe handler are caught and sent to error callback
+PASS inspect(): Provides a way to tap into the moment a source observable is unsubscribed from
+PASS inspect(): Inspector abort() handler is not called if the source completes before the result is unsubscribed from
+FAIL inspect(): Errors thrown from inspect()'s abort() handler are caught and reported to the global, because the subscription is already closed by the time the handler runs self.when is not a function. (In 'self.when('error')', 'self.when' is undefined)
```
Once we implement the [`EventTarget` aspect](https://wicg.github.io/observable/#event-target-integration) of this spec, I believe this tests will start to pass.

I manually tested this particular test with `run-safari --debug`, and can see it working as we expect. 

![image](https://github.com/user-attachments/assets/4d3584bc-46dd-4953-92f3-e0e5c655ca07)

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+
+class InternalObserverInspect final : public InternalObserver {
+public:
+    static Ref<InternalObserverInspect> create(ScriptExecutionContext& context, Ref<Subscriber> subscriber, const ObservableInspector& inspector)
```
`Ref<Subscriber>&&`

maraisr replied:
I could not get the code to compiler with this change. I think given the `handleEvent` is defined like `handleEvent(Subscriber& subscriber)` so, cannot then make a ref out of it? Maybe this line, needs to change?

```cpp
m_sourceObservable->subscribeInternal(*context, InternalObserverInspect::create(*context, subscriber, m_inspector), options);
//                                                                                        ^^^^^^^^^^
```

maraisr replied:
Actually, got it to compile. All good!

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+public:
+    static Ref<InternalObserverInspect> create(ScriptExecutionContext& context, Ref<Subscriber> subscriber, const ObservableInspector& inspector)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverInspect(context, subscriber, inspector));
```
`WTFMove(inspector)`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+
+    class SubscriberCallbackInspect final : public SubscriberCallback {
+    public:
+        static Ref<SubscriberCallbackInspect> create(ScriptExecutionContext& context, Ref<Observable> source, const ObservableInspector& inspector)
```
`Ref<Observable>&&`

cdumez replied:
`ObservableInspector&&` without const.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+    public:
+        static Ref<SubscriberCallbackInspect> create(ScriptExecutionContext& context, Ref<Observable> source, const ObservableInspector& inspector)
+        {
+            return adoptRef(*new InternalObserverInspect::SubscriberCallbackInspect(context, source, inspector));
```
`WTFMove(source)`

cdumez replied:
`WTFMove(inspector)`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+
+                Ref vm = globalObject->vm();
+
+                {
```
Curly brackets here do not seem helpful, it doesn't reduce the scope of the lock in any way.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+    private:
+        bool hasCallback() const final { return true; }
+
+        SubscriberCallbackInspect(ScriptExecutionContext& context, Ref<Observable> source, const ObservableInspector& inspector)
```
`Ref<Observable>&&`

cdumez replied:
`ObservableInspector&&` without const.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+
+        SubscriberCallbackInspect(ScriptExecutionContext& context, Ref<Observable> source, const ObservableInspector& inspector)
+            : SubscriberCallback(&context)
+            , m_sourceObservable(source)
```
`WTFMove(source)`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+            JSC::Exception* exception = scope.exception();
+            if (UNLIKELY(exception)) {
+                scope.clearException();
+                m_subscriber->error(exception->value());
```
Need a smart pointer to m_subscriber *on the stack* before calling the member function on it. The safer cpp bot should complain.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+            }
+        }
+
+        m_subscriber->next(value);
```
Need a smart pointer to m_subscriber *on the stack* before calling the member function on it. The safer cpp bot should complain.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+            JSC::Exception* exception = scope.exception();
+            if (UNLIKELY(exception)) {
+                scope.clearException();
+                m_subscriber->error(exception->value());
```
Need a smart pointer to m_subscriber *on the stack* before calling the member function on it. The safer cpp bot should complain.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+            }
+        }
+
+        m_subscriber->error(value);
```
Need a smart pointer to m_subscriber *on the stack* before calling the member function on it. The safer cpp bot should complain.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+            JSC::Exception* exception = scope.exception();
+            if (UNLIKELY(exception)) {
+                scope.clearException();
+                m_subscriber->error(exception->value());
```
Need a smart pointer to m_subscriber *on the stack* before calling the member function on it. The safer cpp bot should complain.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+            }
+        }
+
+        m_subscriber->complete();
```
Need a smart pointer to m_subscriber *on the stack* before calling the member function on it. The safer cpp bot should complain.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+
+    void visitAdditionalChildren(JSC::SlotVisitor& visitor) const final
+    {
+        m_subscriber->visitAdditionalChildren(visitor);
```
We should avoid code duplication between the 2 visitAdditionalChildren() functions and have them call a templated function (with the visitor type as template parameter).

maraisr replied:
As mentioned over Slack, i've gone with a templated method today, and in a second iteration we will remove the upstream 2nd method, given the AbstractSlotVisitor

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+            return;
+
+        auto handle = std::exchange(m_abortAlgorithmHandler, std::nullopt);
+        m_subscriber->protectedSignal()->removeAlgorithm(*handle);
```
Need a smart pointer to m_subscriber on the stack before calling the member function on it. The safer cpp bot should complain.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+        , m_inspector(inspector)
+    {
+        if (m_inspector.abort) {
+            m_abortAlgorithmHandler = m_subscriber->protectedSignal()->addAlgorithm([abortHandle = Ref { *m_inspector.abort }](JSC::JSValue reason) {
```
`m_subscriber->` -> `subscriber.` to avoid a safer cpp warning.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+        m_subscriber->protectedSignal()->removeAlgorithm(*handle);
+    }
+
+    Ref<JSC::VM> protectedGlobalObjectVM() const
```
This is only used to initialize local variables, it doesn't need to return a smart pointer and doesn't need to have `protected` in the name. I would just use:
`JSC::VM* vm() const;`

Then update the call sites to call `vm()` instead of `protectedGlobalObjectVM()`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+        return globalObject->vm();
+    }
+
+    InternalObserverInspect(ScriptExecutionContext& context, Ref<Subscriber> subscriber, const ObservableInspector& inspector)
```
`Ref<Subscriber>&&`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+
+    InternalObserverInspect(ScriptExecutionContext& context, Ref<Subscriber> subscriber, const ObservableInspector& inspector)
+        : InternalObserver(context)
+        , m_subscriber(subscriber)
```
`WTFMove(subscriber)`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+
+    Ref<Subscriber> m_subscriber;
+    ObservableInspector m_inspector;
+    std::optional<uint32_t> m_abortAlgorithmHandler { };
```
` { }` is unnecessary.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+    std::optional<uint32_t> m_abortAlgorithmHandler { };
+};
+
+Ref<SubscriberCallback> createSubscriberCallbackInspect(ScriptExecutionContext& context, Ref<Observable> observable, RefPtr<JSSubscriptionObserverCallback> next)
```
`Ref<Observable>&&` and `RefPtr<JSSubscriptionObserverCallback>&&`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+
+Ref<SubscriberCallback> createSubscriberCallbackInspect(ScriptExecutionContext& context, Ref<Observable> observable, RefPtr<JSSubscriptionObserverCallback> next)
+{
+    return InternalObserverInspect::SubscriberCallbackInspect::create(context, observable, ObservableInspector { .next = next });
```
`WTFMove(observable)` and `WTFMove(next)`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+    return InternalObserverInspect::SubscriberCallbackInspect::create(context, observable, ObservableInspector { .next = next });
+}
+
+Ref<SubscriberCallback> createSubscriberCallbackInspect(ScriptExecutionContext& context, Ref<Observable> observable, const ObservableInspector& inspector)
```
`Ref<Observable>&&`

cdumez replied:
`ObservableInspector&&` without `const`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+
+Ref<SubscriberCallback> createSubscriberCallbackInspect(ScriptExecutionContext& context, Ref<Observable> observable, const ObservableInspector& inspector)
+{
+    return InternalObserverInspect::SubscriberCallbackInspect::create(context, observable, inspector);
```
`WTFMove(observervable)`

cdumez replied:
`WTFMove(inspector)`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.h:
```diff
@@ -0,0 +1,41 @@
+class SubscriberCallback;
+struct ObservableInspector;
+
+Ref<SubscriberCallback> createSubscriberCallbackInspect(ScriptExecutionContext&, Ref<Observable>, RefPtr<JSSubscriptionObserverCallback>);
```
`&&` for smart pointers

cdumez commented on Source/WebCore/dom/InternalObserverInspect.h:
```diff
@@ -0,0 +1,41 @@
+struct ObservableInspector;
+
+Ref<SubscriberCallback> createSubscriberCallbackInspect(ScriptExecutionContext&, Ref<Observable>, RefPtr<JSSubscriptionObserverCallback>);
+Ref<SubscriberCallback> createSubscriberCallbackInspect(ScriptExecutionContext&, Ref<Observable>, const ObservableInspector&);
```
`Ref<Observable>&&`

cdumez replied:
`ObservableInspector&&` without `const`

cdumez commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -122,6 +124,22 @@ Ref<Observable> Observable::drop(ScriptExecutionContext& context, uint64_t amoun
+        return *this;
+
+    return WTF::switchOn(
+        inspectorUnion.value(),
```
`WTFMove(*inspectorUnion)`

maraisr replied:
This seems to fail with:

```
WebKit/WebKitBuild/Debug/usr/local/include/wtf/StdLibExtras.h:531:9: error: no matching function for call to 'invoke'
  531 |         WTF_VISIT_CASE(1, WTF_VISIT_CASE_COUNT)
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
WebKit/WebKitBuild/Debug/usr/local/include/wtf/StdLibExtras.h:523:24: note: expanded from macro 'WTF_VISIT_CASE'
  523 |                 return std::invoke(std::forward<F>(f), std::get<I + N>(std::forward<V>(v)));        \
```

cdumez commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -122,6 +124,22 @@ Ref<Observable> Observable::drop(ScriptExecutionContext& context, uint64_t amoun
+
+    return WTF::switchOn(
+        inspectorUnion.value(),
+        [&](RefPtr<JSSubscriptionObserverCallback>& next) {
```
`RefPtr<JSSubscriptionObserverCallback>&&`

cdumez commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -122,6 +124,22 @@ Ref<Observable> Observable::drop(ScriptExecutionContext& context, uint64_t amoun
+    return WTF::switchOn(
+        inspectorUnion.value(),
+        [&](RefPtr<JSSubscriptionObserverCallback>& next) {
+            return create(createSubscriberCallbackInspect(context, *this, next));
```
`WTFMove(next)`

cdumez commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -122,6 +124,22 @@ Ref<Observable> Observable::drop(ScriptExecutionContext& context, uint64_t amoun
+        [&](RefPtr<JSSubscriptionObserverCallback>& next) {
+            return create(createSubscriberCallbackInspect(context, *this, next));
+        },
+        [&](ObservableInspector& inspector) {
```
`ObservableInspector&&`

maraisr replied:
Doing this, created a similar error to [this](https://github.com/WebKit/WebKit/pull/37520#discussion_r1874197025)

cdumez commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -122,6 +124,22 @@ Ref<Observable> Observable::drop(ScriptExecutionContext& context, uint64_t amoun
+            return create(createSubscriberCallbackInspect(context, *this, next));
+        },
+        [&](ObservableInspector& inspector) {
+            return create(createSubscriberCallbackInspect(context, *this, inspector));
```
`WTFMove(inspector)`

cdumez commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -122,6 +124,22 @@ Ref<Observable> Observable::drop(ScriptExecutionContext& context, uint64_t amoun
     return create(createSubscriberCallbackDrop(context, *this, amount));
 }
 
+Ref<Observable> Observable::inspect(ScriptExecutionContext& context, std::optional<InspectorUnion> inspectorUnion)
```
`std::optional<InspectorUnion>&&`

maraisr replied:
As mentioned [here](https://github.com/WebKit/WebKit/pull/37520#discussion_r1874197025) I was not able to get a compilation doing this.

cdumez commented on Source/WebCore/dom/Observable.h:
```diff
@@ -57,12 +59,10 @@ class Observable final : public ScriptWrappable, public RefCounted<Observable> {
     Ref<Observable> take(ScriptExecutionContext&, uint64_t);
-
     Ref<Observable> drop(ScriptExecutionContext&, uint64_t);
+    Ref<Observable> inspect(ScriptExecutionContext&, std::optional<InspectorUnion>);
```
`std::optional<InspectorUnion>&&`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+        SubscriberCallbackInspect(ScriptExecutionContext& context, Ref<Observable> source, const ObservableInspector& inspector)
+            : SubscriberCallback(&context)
+            , m_sourceObservable(source)
+            , m_inspector(inspector)
```
`WTFMove(inspector)`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+                return { };
+            }
+
+            if (m_inspector.subscribe) {
```
`if (RefPtr subscribe = m_inspector.subscribe) {`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+                JSC::JSLockHolder lock(vm);
+                auto scope = DECLARE_CATCH_SCOPE(vm);
+
+                m_inspector.subscribe->handleEventRethrowingException();
```
`subscribe->handleEventRethrowingException();`

Per our safer C++ guidelines, you need a smart pointer *on the stack* before calling a non-trivial member function.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+                }
+            }
+
+            SubscribeOptions options;
```
`SubscribeOptions options { &subscriber.signal() };`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+            }
+
+            SubscribeOptions options;
+            options.signal = &subscriber.signal();
```
Then we don't need this.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+
+            SubscribeOptions options;
+            options.signal = &subscriber.signal();
+            m_sourceObservable->subscribeInternal(*context, InternalObserverInspect::create(*context, subscriber, WTFMove(m_inspector)), options);
```
`Ref { m_sourceObservable }->` per safer C++ guidelines.

cdumez replied:
`WTFMove(m_inspector)` seems wrong here, you likely want `ObservableInspector { m_inspector }`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+
+        removeAbortHandler();
+
+        if (m_inspector.complete) {
```
`if (RefPtr complete = m_inspector.complete) {`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+            JSC::JSLockHolder lock(vm);
+            auto scope = DECLARE_CATCH_SCOPE(vm);
+
+            m_inspector.complete->handleEventRethrowingException();
```
`complete->handleEventRethrowingException();`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+        protectedSubscriber()->protectedSignal()->removeAlgorithm(*handle);
+    }
+
+    Ref<JSC::VM> protectedVM() const
```
This should be called `vm()` and return a `JSC::vm*`, as mentioned in an earlier review. It is only use to initialize local variables.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+    void next(JSC::JSValue value) final
+    {
+        if (m_inspector.next) {
+            Ref vm = this->protectedVM();
```
`Ref vm = this->vm();`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+    {
+        removeAbortHandler();
+
+        if (m_inspector.error) {
```
`if (RefPtr error = m_inspector.error) {`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+            JSC::JSLockHolder lock(vm);
+            auto scope = DECLARE_CATCH_SCOPE(vm);
+
+            m_inspector.error->handleEventRethrowingException(value);
```
`error->handleEventRethrowingException(value);`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+        , m_inspector(WTFMove(inspector))
+    {
+        if (m_inspector.abort) {
+            Ref signal = protectedSubscriber()->protectedSignal();
```
`Ref signal = protectedSubscriber()->signal();`

`protectedFoo()` functions are only used to avoid local variables, not to initialize them.

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+private:
+    void next(JSC::JSValue value) final
+    {
+        if (m_inspector.next) {
```
`if (RefPtr next = m_inspector.next) {`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,252 @@
+            JSC::JSLockHolder lock(vm);
+            auto scope = DECLARE_CATCH_SCOPE(vm);
+
+            m_inspector.next->handleEventRethrowingException(value);
```
`next->handleEventRethrowingException(value);`

cdumez commented on Source/WebCore/dom/InternalObserverInspect.cpp:
```diff
@@ -0,0 +1,253 @@
+                }
+            }
+
+            Ref { m_sourceObservable }->subscribeInternal(*context,
```
nit: This line is very long. Maybe we can use a local to make it shorter:
```
Ref inspect = InternalObserverInspect::create(*context, subscriber, ObservableInspector { m_inspector });
Ref { m_sourceObservable }->subscribeInternal(*context, WTFMove(inspect), SubscribeOptions { &subscriber.signal() });
```

cdumez reviewed (approved):
LGTM

maraisr committed:
Implement Observable#inspect
https://bugs.webkit.org/show_bug.cgi?id=282083

Reviewed by Chris Dumez.

This change introduces the .inspect operator for Observables, using the
InternalObserver, and a new SubscriberCallbackInspect. .inspect is meant to act
as a pass-through operator, so there are no mutations on events.

Spec: <https://wicg.github.io/observable/#dom-observable-inspect>

There is also a new ObservableInspector IDL dictionary created, as well as a
ObservableInspectorAbortCallback. We could not just have used some generic
callback IDL, given we require a callback that accepts a JSC::Value, as we
forward abort reasons to this callback.

* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-inspect.any-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-inspect.any.worker-expected.txt: Rebaseline.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/InternalObserverInspect.cpp: Added.
(WebCore::createSubscriberCallbackInspect):
* Source/WebCore/dom/InternalObserverInspect.h: Added.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::inspect):
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Observable.idl:
Exposes a `Observable inspect(inspectorUnion)` operator.
* Source/WebCore/dom/ObservableInspector.h: Added.
* Source/WebCore/dom/ObservableInspector.idl: Added.
* Source/WebCore/dom/ObservableInspectorAbortCallback.h: Added.
* Source/WebCore/dom/ObservableInspectorAbortCallback.idl: Added.

Canonical link: https://commits.webkit.org/287504@main
